### PR TITLE
Add config option for state_store.watchLimit

### PR DESCRIPTION
**Pull request by ShimmerGlass**
As described here : https://github.com/hashicorp/consul/issues/4984, exceeding watchLimit greatly degrades blocking queries performance. 
This patch allows to configure this parameter while keeping the current value of 2048 as default, and displays a warning when the value is reached.

### CODE REVIEW

File: agent/config/builder.go [hunk 1]
```diff
@@ -843,6 +843,7 @@ func (b *Builder) Build() (rt RuntimeConfig, err error) {
 		VerifyOutgoing:                          b.boolVal(c.VerifyOutgoing),
 		VerifyServerHostname:                    b.boolVal(c.VerifyServerHostname),
 		Watches:                                 c.Watches,
+		WatchSoftLimit:                          b.intValWithDefault(c.WatchSoftLimit, 2048),
```
Comment (pierresouchay) [inline agent/config/builder.go]: Can you use the constant watchLimit instead?

File: agent/config/runtime_test.go [hunk 2]
```diff
@@ -3893,7 +3894,8 @@ func TestFullConfig(t *testing.T) {
 				datacenter = "fYrl3F5d"
 				key = "sl3Dffu7"
 				args = ["dltjDJ2a", "flEa7C2d"]
-			}]
+			}],
+			watch_soft_limit = 2048
```
Comment (pierresouchay) [inline agent/config/runtime_test.go]: same, if possible use the constant

File: agent/config/runtime_test.go [hunk 3]
```diff
@@ -3347,7 +3347,8 @@ func TestFullConfig(t *testing.T) {
 					"key": "sl3Dffu7",
 					"args": ["dltjDJ2a", "flEa7C2d"]
 				}
-			]
+			],
+			"watch_soft_limit": 2048
```
Comment (pierresouchay) [inline agent/config/runtime_test.go]: constant

File: agent/config/runtime.go [hunk 4]
```diff
@@ -1413,6 +1413,14 @@ type RuntimeConfig struct {
 	// ]
 	//
 	Watches []map[string]interface{}
+
+	// WatchSoftLimit is used as a soft limit to cap how many watches we allow
+	// for a given blocking query. If this is exceeded, then we will use a
+	// higher-level watch that's less fine-grained.
+	// Default to 2048
```
Comment (pierresouchay) [inline agent/config/runtime.go]: use constant

File: agent/config/runtime_test.go [hunk 5]
```diff
@@ -4538,6 +4540,7 @@ func TestFullConfig(t *testing.T) {
 				"args":       []interface{}{"dltjDJ2a", "flEa7C2d"},
 			},
 		},
+		WatchSoftLimit: 2048,
```
Comment (pierresouchay) [inline agent/config/runtime_test.go]: use watchLimit

**Comment by Aestek** [comment]
Sure @banks !

**Comment by Aestek** [comment]
@banks done

**Comment by pierresouchay** [comment]
I think @orarnon has the same issue as we did:

he creates a `/etc/hosts` file using consul-template with all the nodes with more than 2048 nodes => performance drop when services do flap. Shutting down consul-template solves the performance.

**Comment by orarnon** [comment]
@pierresouchay Yes, that's exactly right. 
Since we render all hosts into our hosts file, we have over 2K nodes at all times which degregates performance. 
When we add services flapping, it practically kills Consul cluster. Our only mitigation was to stop Consul-Template on all hosts and let things calm down.
Per Pierre's advice, we'll begin with listing only our data services in our hosts file which will create several watchers but with 5-20 instances each. So we expect much smaller hosts file with less events.

**Comment by banks** [comment]
Thanks for the updates Or Arnon.

For the record we held out on this PR as we are getting closer to replacing
the watching mechanism entirely with a much much more efficient version in
the near future. We'll note that here when it is ready (hopefully soon!).

Paul

**Comment by i0rek** [comment]
Is this still something that you need?

**Comment by pierresouchay** [comment]
@i0rek We are using this with value 8192 for more than 1y... it solved most of our issues on large services...

Several people had this issue has well , including @orarnon  and @yiliaofan 
For what we saw (and until streaming is available, this is one of the most important patch for large infrastructures and having a greater limit than 2018 by default would avoid many issues).

An article describing the issue we had: https://medium.com/criteo-labs/anatomy-of-a-bug-when-consul-has-too-much-to-deliver-for-the-big-day-4904d19a46a4

**Comment by orarnon** [comment]
as @pierresouchay stated, we have encountered this issue multiple times and it killed our production environment more than once as we rely on Consul for so many things.

**Comment by banks** [comment]
To fill in the history, this PR predates a bunch of optimization work we did as part of the issue this references.

> For what we saw (and until streaming is available, this is one of the most important patch for large infrastructures

We shipped a fix that (as far as I understood) made this optimization unnecessary at least for health/service watches in Consul 1.4.4 in March 2019.

@pierresouchay does this patch still make a difference on your clusters since 1.4.4. optimisation that [makes all service health queries a since watch chan anyway](https://github.com/hashicorp/consul/pull/5449)? Were you also exceeding this limit in other types of blocking query?

I thought the only reason we kept the original issue open was to track the more advanced fixes which will eventually be streaming?

I'd still rather ship streaming and improve this issue by orders of magnitude that ship yet another tunable that operators shouldn't need to have to touch but if this is actually still necessary even in 1.4.4+ that would be good to know.

**Comment by pierresouchay** [comment]
@banks We still do use the blocking queries patches and RPC (but we plan to leave it as soon as streaming can be validated on our side). However, some people, such as @orarnon had issues with 2048 limit on catalog/nodes for instance (using consul-template watch to generate host files if I am not mistaken).

Not leaving the 2048 value configurable is perfectly understandable, however, increasing it to a more important value (with use 8192 for months now without any issue because it suits us with our current loads, but probably, removing it completely would be even better) might make sense until streaming comes to mainstream (I know 2 different people from 2 companies that had this issue, so this is just the tip of iceberg). Maybe at least providing a log when the limit is reached would give more insights about the breakage it creates in the real world.

To me, the value 2048 is just a defensive programing practice (I would have done similar thing naturally) that does not stand the real work scenario has it breaks very quickly when it reaches the limit anyway (the remedy is worse than what it tries to avoid aka multiplying too much the number of go-routines because the O(n) explodes when the limit is reached)

**Comment by orarnon** [comment]
Hi,
@pierresouchay is right. Since there's no documentation or logs on this matter, we were puzzled for weeks as to why our Consul cluster was not handling the load despite the fact that it was over-provisioned and optimized. It caused us multiple downtimes to our production environment to the point where we considered replacing Consul with another solution.
After a short conversation with Pierre, where he explained this issue to me, we have reduced out hosts file to show data services only. This change reduced the number of hosts below the 2048 threshold.

Since this change, we did not even have a leader change and could sustain volatile changes in our environment without killing Consul.
@banks I urge you to find a streaming solution one way or another. Personally, I explained this issue to a couple of companies as well and helped them solved that mystery.

**Comment by robloxrob** [comment]
We are experiencing this issue and are working with HashiCorp support on the matter. Given that this issue seems so prevalent at scale adding some documentation, notes or warnings would be useful for future visitors who could run into this issue and come visit this issue.

**Comment by banks** [comment]
@pierresouchay @orarnon @robloxrob, thanks for this context - that's really useful info.

We'll talk about the best path forward. I'm leaning towards just changing the limit to something much higher as an interim solution until streaming is available everywhere given how likely it is that if you have large results, you also have lots of churn which is what makes the fallback to the table index way more costly in these examples.

I'll leave this open until we've decided which way to go.

**Comment by robloxrob** [comment]
@banks Thank you for your consideration in this matter.

**Comment by i0rek** [comment]
Thank you for helping us understand the importance of this value to be higher. We don't want to add another option for it because it is just another thing to fine tune and because we assume streaming will fix it. But we went ahead and increased it to what @pierresouchay has been suggesting: `8192`: https://github.com/hashicorp/consul/pull/7200.

This is the reason I am closing this PR. Thanks again!

**Pull request closed without merging**